### ISSUE TITLE
References to type synonyms for promoted data constructors instead link to the data type

### ISSUE THREAD

**Issue by rightfold** [haskell/haddock#1047]
Consider the following Haskell file:

```haskell
{-# LANGUAGE DataKinds #-}

module Example
  ( Bla (..)
  , Foo
  , Bar
  ) where

-- |
-- 'Bla' 'Foo' 'Bar'.
data Bla
  = Foo
  | Bar
type Foo = 'Foo
type Bar = 'Bar
```

Haddock renders the documentation for Bla as three links to Bla, instead of linking to Bla, Foo, and Bar as appropriate. I.e. the documentation shows up as:

> Bla Bla Bla.

Using Haddock 2.22.0.

**Comment by harpocrates** [comment]
This is indeed problematic. Another issue with the generated docs for your code sample: `type Foo = Foo` is not very helpful at all! We should try to render `type Foo  = 'Foo` if at all possible.

**Comment by harpocrates** [comment]
Turns out the quote issue I was worried about is already fixed in 7dd0a79cce7c4c048e7c145c9f378da3a96392d0. The original bug stands.

**Comment by harpocrates** [comment]
Turns out this is a duplicate of #1035, already fixed in b682041ed1cbeaf5aa501f85e4e46a6d2e39da3a (which will be in the upcoming Haddock 2.23 release, bundled with GHC 8.8.1). Visual confirmation:

<img width="428" alt="Screen Shot 2019-05-28 at 3 37 43 PM" src="https://user-images.githubusercontent.com/10766081/58506944-951b1600-815e-11e9-8a82-02fe53d28193.png">
